File: src/master-data-reconciliation/src/main/java/io/ioka/demo/reconcile/Bootstrap.java
// SPDX-License-Identifier: Apache-2.0
package io.ioka.demo.reconcile;

import io.ioka.munarium.client.*;
import io.ioka.munarium.client.model.*;
import java.nio.file.*;
import java.util.*;

public final class Bootstrap {
    private Bootstrap() {}
    public static final String REVISION = "bb6e92a72a3944cff4d4bf0c1b470afcf3f4dfb3";
    public record Grant(String token, String uid, String namespace, String provider, String model, String config, Map<String,String> runbooks) {}
    public static String endpoint() { return FilesUtil.env("MUNARIUM_REST_URL", "http://server:8080"); }
    public static MunariumClient client(String token, String uid) {
        return MunariumClient.rest(MunariumClientOptions.of(endpoint()).withToken(token).withUid(uid).withReadRetries(0));
    }
    public static MunariumClient ops(boolean management) {
        String token = System.getenv(management ? "MUNARIUM_MGMT_TOKEN" : "MUNARIUM_TOKEN");
        if (token == null) throw new IllegalStateException("Operator credential is absent.");
        return client(token, "reconcile-bootstrap");
    }
    public static Grant grant() throws Exception { return Json.MAPPER.readValue(Files.readString(Path.of("/credentials/query.json")), Grant.class); }
    public static void run(String provider, boolean approve) throws Exception {
        Fixtures.assignments(provider);
        if (!approve) throw new IllegalArgumentException("Bootstrap requires --approve for its isolated test index cutovers.");
        Path inputs = Path.of("/inputs"); Fixtures.verify(inputs);
        if (!provider.equals("fixture") && !FilesUtil.read(inputs.resolve("manifest.json")).path("profile").asText().equals("default")) throw new IllegalArgumentException("Online qualification requires default fixtures");
        String model = provider.equals("fixture") ? "reconcile-selected" : System.getenv(provider.toUpperCase(Locale.ROOT) + "_MODEL");
        if (model == null || model.isBlank()) throw new IllegalArgumentException("Preferred model is absent.");
        String template = Files.readString(Path.of("/app/runbooks/stewardship.yaml"));
        String shape = Files.readString(Path.of("/app/shapes/documents.yaml"));
        String name = "reconcile-" + FilesUtil.hash(Files.readString(inputs.resolve("manifest.json")) + template + shape + Files.readString(Path.of("/app/shapes/claims.yaml")) + REVISION + provider + model).substring(0, 12);
        String config = name + "-model";
        Map<String,String> refs = new TreeMap<>();
        Map<String,String> runs = new TreeMap<>();
        try (var ops = ops(false)) {
            for (int attempt = 0; ; attempt++) {
                try {
                    if (!ops.serverVersion().version().equals("1.1.1")) throw new IllegalStateException("Server 1.1.1 required.");
                    break;
                } catch (io.ioka.munarium.client.errors.MunariumException e) { if (attempt == 59) throw e; Thread.sleep(2000); }
            }
            String connection = provider.equals("fixture") ? "endpoint: http://provider-fixture:11434" : "credentialRef: {env: " + provider.toUpperCase(Locale.ROOT) + "_API_KEY}";
            ops.providers.applyConfig("""
                apiVersion: munarium.ioka.io/v1
                kind: ProviderConfig
                metadata: {name: %s}
                spec:
                  provider: %s
                  %s
                  models:
                    complete: [%s]
                    fast: %s
                  budgets: {rpm: %d, dailyTokens: {fast: 100000}}
                """.formatted(config, provider.equals("fixture") ? "ollama" : provider, connection, FilesUtil.json(model), FilesUtil.json(model), provider.equals("fixture") ? Math.max(60, Fixtures.rows(inputs).size() * 3) : 60));
            if (!ops.providers.health(config).healthy()) throw new IllegalStateException("Named provider health failed: " + provider);
            ops.runbooks.applyShape(shape, null);
            ops.runbooks.applyShape(Files.readString(Path.of("/app/shapes/claims.yaml")), null);
            for (var event : Fixtures.rows(inputs)) {
                String scoped = name + "-" + event.id();
                ops.runbooks.applyRunbook(template.replace("__NAME__", scoped).replace("__PROVIDER__", config));
                var result = ops.ingest.ingest(Ingesting.IngestFile.ofText(scoped + "/procedure.txt", "text/plain", Files.readString(inputs.resolve("documents/" + event.id() + ".txt"))));
                if (result.error() != null || !result.boundTo().contains(scoped)) throw new IllegalStateException("Unexpected ingest binding.");
                var run = ops.runbooks.runRunbook(scoped, null);
                runs.put(event.id(), run.runId());
                FilesUtil.save(Path.of("/work/bootstrap/" + provider + "-" + run.runId() + ".json"), Map.of("runs", runs, "client_revision", REVISION, "namespace", name, "provider", provider, "model", model, "manifest", FilesUtil.read(inputs.resolve("manifest.json"))));
                var pending = ops.runbooks.getRun(run.runId()).steps().stream().filter(s -> s.state().equals("awaiting_approval")).toList();
                if (pending.size() != 1 || !pending.getFirst().name().equals("cutover:" + scoped)) throw new IllegalStateException("Unexpected index cutover.");
                ops.runbooks.approveStep(run.runId(), pending.getFirst().ordinal());
                if (!ops.runbooks.getRun(run.runId()).state().equals("done")) throw new IllegalStateException("Index run incomplete.");
                refs.put(event.id(), scoped + "@1");
            }
        }
        try (var issuer = ops(true)) {
            var issued = issuer.tokens.mint(new Tokens.IssueTokenRequest("reconcile-reviewer", 0, List.of(), List.of("query"), refs.values().stream().map(r -> r.split("@")[0]).toList(), 3600L));
            FilesUtil.save(Path.of("/credentials/query.json"), new Grant(issued.token(), "reconcile-reviewer", name, provider, model, config, refs));
        }
        System.out.println("Bootstrapped " + provider + "/" + model + "; " + refs.size() + " isolated stewardship collections.");
    }
}

File: src/maintenance-terminal/src/bootstrap.rs
// SPDX-License-Identifier: Apache-2.0
use crate::data::*;
use anyhow::{ensure, Result};
use base64::{engine::general_purpose::STANDARD, Engine};
use munarium_client::{dto, MunariumClient, MunariumClientOptions};
use serde::{Deserialize, Serialize};
use serde_json::json;
use std::{collections::BTreeMap, env, fs, time::Duration};

#[derive(Clone, Serialize, Deserialize)]
pub struct Scope {
    pub retrieve: String,
    pub explain: String,
    pub collection: String,
}
#[derive(Clone, Serialize, Deserialize)]
pub struct Grant {
    pub token: String,
    pub uid: String,
    pub provider: String,
    pub model: String,
    pub config: String,
    pub scopes: BTreeMap<String, Scope>,
}
pub fn endpoint() -> String {
    env::var("MUNARIUM_REST_URL").unwrap_or("http://server:8080".into())
}
pub fn client(token: &str, uid: &str) -> Result<MunariumClient> {
    client_at(&endpoint(), token, uid)
}
pub fn client_at(url: &str, token: &str, uid: &str) -> Result<MunariumClient> {
    Ok(MunariumClient::rest(
        MunariumClientOptions::new(url)
            .token(token)
            .uid(uid)
            .read_retries(0),
    )?)
}
pub fn ops(management: bool) -> Result<MunariumClient> {
    client(
        &env::var(if management {
            "MUNARIUM_MGMT_TOKEN"
        } else {
            "MUNARIUM_TOKEN"
        })?,
        "maintenance-bootstrap",
    )
}
pub fn grant() -> Result<Grant> {
    Ok(serde_json::from_value(read("/credentials/query.json")?)?)
}
pub async fn ready(c: &MunariumClient) -> Result<()> {
    for _ in 0..60 {
        if let Ok(v) = c.server_version().await {
            ensure!(v.version == "1.1.1", "Server 1.1.1 required");
            return Ok(());
        }
        tokio::time::sleep(Duration::from_secs(2)).await;
    }
    anyhow::bail!("Server did not become ready")
}
pub async fn mint(uid: &str, refs: Vec<String>, ttl: u64) -> Result<String> {
    Ok(ops(true)?
        .tokens
        .mint(dto::IssueTokenRequest {
            uid: uid.into(),
            access_level: 0,
            compartments: vec![],
            scopes: vec!["query".into()],
            runbook_refs: Some(refs),
            ttl_secs: Some(ttl),
        })
        .await?
        .token)
}
pub async fn run(provider: &str, approve: bool) -> Result<()> {
    ensure!(
        approve,
        "Bootstrap requires --approve for this isolated demo's cutovers"
    );
    assignments(provider)?;
    verify("/inputs")?;
    ensure!(
        provider == "fixture" || read("/inputs/manifest.json")?["profile"] == "default",
        "Online qualification requires default fixtures"
    );
    let model = if provider == "fixture" {
        "maintenance-selected".into()
    } else {
        env::var(format!("{}_MODEL", provider.to_uppercase()))?
    };
    ensure!(!model.trim().is_empty(), "Model is required");
    let template = fs::read_to_string("runbooks/maintenance.yaml")?;
    let shape = fs::read_to_string("shapes/documents.yaml")?;
    let completion = fs::read_to_string("runbooks/completion.yaml")?;
    let namespace = format!(
        "maint-{}",
        &hash(format!(
            "{}{}{}{}{}{}{}",
            fs::read_to_string("/inputs/manifest.json")?,
            template,
            shape,
            completion,
            REVISION,
            provider,
            model
        ))[..12]
    );
    let config = format!("{namespace}-model");
    let c = ops(false)?;
    ready(&c).await?;
    let connection = if provider == "fixture" {
        "endpoint: http://provider-fixture:11434".into()
    } else {
        format!(
            "credentialRef: {{env: {}_API_KEY}}",
            provider.to_uppercase()
        )
    };
    let family = if provider == "fixture" {
        "ollama"
    } else {
        provider
    };
    let model_json = serde_json::to_string(&model)?;
    let rpm = if provider == "fixture" {
        (catalogue("/inputs")?.len() * 3).max(30)
    } else {
        30
    };
    c.providers.apply_config(&format!("apiVersion: munarium.ioka.io/v1\nkind: ProviderConfig\nmetadata: {{name: {config}}}\nspec:\n  provider: {family}\n  {connection}\n  models:\n    complete: [{model_json}]\n    fast: {model_json}\n  budgets: {{rpm: {rpm}, dailyTokens: {{fast: 100000}}}}\n")).await?;
    ensure!(
        c.providers.health(&config).await?.healthy,
        "Named provider health failed: {provider}"
    );
    c.runbooks.apply_shape(&shape, None).await?;
    let mut scopes = BTreeMap::new();
    for asset in catalogue("/inputs")? {
        let collection = format!("{namespace}-{}", asset.id);
        let retrieve = format!("{collection}-find");
        let explain = format!("{collection}-explain");
        for (name, completion) in [
            (&retrieve, String::new()),
            (&explain, completion.replace("__PROVIDER__", &config)),
        ] {
            c.runbooks
                .apply_runbook(
                    &template
                        .replace("__NAME__", name)
                        .replace("__COLLECTION__", &collection)
                        .replace("__COMPLETION__", &completion),
                )
                .await?;
        }
        for kind in ["manual", "inspection", "revision"] {
            let content = fs::read(format!("/inputs/documents/{}-{kind}.txt", asset.id))?;
            let result = c
                .ingest
                .ingest(dto::IngestFileRequest {
                    filename: format!("{collection}/{kind}.txt"),
                    media_type: "text/plain".into(),
                    content_base64: STANDARD.encode(content),
                    sha256: None,
                    collections: None,
                })
                .await?;
            ensure!(
                result.error.is_none() && result.bound_to.contains(&collection),
                "Unexpected ingest binding"
            );
        }
        let run = c.runbooks.run_runbook(&retrieve, None).await?;
        save(
            format!("/work/bootstrap/{provider}-{}.json", run.run_id),
            &json!({"run_id":run.run_id,"collection":collection,"manifest":read("/inputs/manifest.json")?,"client_revision":REVISION,"provider":provider,"model":model}),
        )?;
        let state = c.runbooks.get_run(&run.run_id).await?;
        let pending: Vec<_> = state
            .steps
            .iter()
            .filter(|s| s.state == "awaiting_approval")
            .collect();
        ensure!(
            pending.len() == 1 && pending[0].name == format!("cutover:{collection}"),
            "Unexpected cutover"
        );
        c.runbooks
            .approve_step(&run.run_id, pending[0].ordinal)
            .await?;
        ensure!(
            c.runbooks.get_run(&run.run_id).await?.state == "done",
            "Incomplete index run"
        );
        scopes.insert(
            asset.id,
            Scope {
                retrieve: format!("{retrieve}@1"),
                explain: format!("{explain}@1"),
                collection,
            },
        );
    }
    let refs = scopes
        .values()
        .flat_map(|s| {
            [
                s.retrieve.split('@').next().unwrap().into(),
                s.explain.split('@').next().unwrap().into(),
            ]
        })
        .collect();
    let token = mint("maintenance-reader", refs, 3600).await?;
    save(
        "/credentials/query.json",
        &Grant {
            token,
            uid: "maintenance-reader".into(),
            provider: provider.into(),
            model,
            config,
            scopes,
        },
    )?;
    println!("Ready: {provider}; selected asset/revision scopes; query capability issued.");
    Ok(())
}
